lib/cli/src/package_source.rs

```diff
@@ -0,0 +1,188 @@
+//! Module for parsing and installing packages
+
+use anyhow::Context;
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+use url::Url;
+
+/// Source of a package
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum PackageSource {
+    /// Download from a URL
+    Url(Url),
+    /// Run a local file
+    File(String),
+    /// Download from a package
+    Package(wasmer_registry::Package),
+}
+
+impl Default for PackageSource {
+    fn default() -> Self {
+        PackageSource::File(String::new())
+    }
+}
+
+impl FromStr for PackageSource {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Self::parse(s)
+    }
+}
+
+impl PackageSource {
+    /// Parses a package source and transforms it to a URL or a File
+    pub fn parse(s: &str) -> Result<Self, String> {
+        // If the file is a http:// URL, run the URL
+        if let Ok(url) = url::Url::parse(s) {
+            if url.scheme() == "http" || url.scheme() == "https" {
+                return Ok(Self::Url(url));
```

Comment:
We probably don't want to accept just any URL, but only urls that conform to the package spec.

Just requires checking the `url.path().strip_prefix('/')` with `Package::from_str()` and report something like an "invalid URL" error.

Reply:
We want to only accept `http://` URLs because the other Urls were conflicting with Windows' `C://` formatting (the original source of the bug that started this whole PR).

Reply:
If you mean, only accepting very narrow `https://wapm.dev/namespace/package` and for example, not `https://wapm.dev/bb0df255edf049905a5d23b03e59a0c240d9723c`, that might cause unexpected issues later on. I would not do this.

Reply:
That's a valid argument.

It's a better UX if we get a sensible error, because we'll later get something like a `could not parse ` error when you use a URL that doesn't return a webc.

But that's probably fine.

Reply:
@theduke If the URL doesn't return a webc file, it will print the first 100 bytes:

https://github.com/wasmerio/wasmer/pull/3416/files#diff-d83774b7aa5a38e272bd0772c3f95c406608bf13c31c699652c9ce0356c2e2a1R1192-R1198